src/processor/mod.rs

```diff
@@ -0,0 +1,210 @@
+use std::convert::Infallible;
+use std::ffi::OsStr;
+use std::fmt::{Display, Formatter};
+use std::fs::File;
+use std::io::{BufReader, Error, Read, Write};
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+/// Represents a type that can process a given file for the web.
+trait Processor {
+    /// Creates a new processor for the given file.
+    ///
+    /// # Arguments
+    ///
+    /// * `file`: The file to process.
+    ///
+    /// returns: A Processor implementation.
+    fn new(file: File) -> Self where Self: Sized;
+
+    /// A want-to-be field for the type of processor.
+    ///
+    /// Ideally, this would be a simple property, as the look- up time is
+    /// constant. However, Rust does not yet support trait fields. You can read
+    /// more about that work at: https://github.com/rust-lang/rfcs/pull/1546
+    ///
+    /// It is also highly likely that I don't know what I'm doing, and there is
+    /// a better way.
+    ///
+    /// returns: The concrete type of processor.
+    fn kind(&self) -> ProcessorKind;
+
+    /// Performs the processing of the file and writes the result.
+    ///
+    /// # Arguments
+    ///
+    /// * `path`: Where the results will be written.
+    ///
+    /// returns: A result indicating the outcome.
+    fn write_to(&mut self, path: &Path) -> Result<(), ProcessorError>;
+}
+
+struct ProcessorError {
+    path: String,
+}
+
+impl From<std::io::Error> for ProcessorError {
+    fn from(_: Error) -> Self {
+        // ideally, would get the path here (or perhaps should convert)
+       ProcessorError { path: String::from("(unknown)") }
+    }
+}
+
+impl Display for ProcessorError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ProcessorError {{ path: {} }}", self.path)
+    }
+}
+
+/// The type of file we are processing (named after the extension).
+#[derive(Debug, PartialEq)]
+enum ProcessorKind {
+    None,
+    Css,
+    Md,
+    Org,
+}
+
+impl FromStr for ProcessorKind {
```

Comment:
```suggestion
impl From<&str> for ProcessorKind {
```

Reply:
`FromStr` is great for parsing stuff because it can fail, but that's not what's here. (Typing other's comments makes me look smart.)